### ISSUE TITLE
OpenVINO outputs different results under Inter CPU and Inter GPU

### ISSUE THREAD

**Issue by fanxiaochen456** [openvinotoolkit/openvino#12096]
Hi, I used the script of MO_tf.py to convert the Tensorflow solidified model with the accuracy of FP32 and the PB file to the IR model of FP16. The detection result under Inter CPU is correct, but when the Device is replaced with Inter GPU, The reasoning result of the model is different from the correct output result under CPU, and there is no error with custom layer when converting IR model. What is the reason?

**Comment by fanxiaochen456** [comment]
I looked at the XML files converted by FP32 and FP16, and the network structure was no different. The input and output dimensions are also normal.

**Comment by jgespino** [comment]
@fanxiaochen456 Could you please share additional information to reporduce from our end?

- OpenVINO Version
- Model
- Model Optimzier Command
- Test input
- Expected output 
- Code to reproduce

**Comment by fanxiaochen456** [comment]
@jgespinoHello, THE OpenVINO version I'm using is OpenVino_2020.3.355 +VS2017+Python3.6.8. The OpenVINO environment configured is as follows:
Package                       Version
----------------------------- ---------
Absl - py 0.15.0
Astor 0.8.0
Astunparse 1.6.3
Cachetools 4.2.4
Certifi 2022.6.15
Charset - normalizer 2.0.12
Cycler 0.10.0
Dataclasses 0.8
The decorator 4.4.2
Defusedxml 0.7.1
Enum34 1.1.6
Flatbuffers 1.12
Gast 0.3.3
Google - auth 2.9.0
Google - auth - oauthlib 0.4.6
Google - pasta 0.2.0
Graphviz 0.19.1
Grpcio 1.32.0
H5py 2.10.0
Idna 3.3
Keras 2.3.0
Keras - Applications 1.0.8
Keras - Preprocessing 1.1.2
Kiwisolver 1.1.0
Markdown 3.1.1
Matplotlib 3.1.1
Matplotlib - Venn 0.11.7
Mxnet 1.2.0
Networkx 2.5.1
Numpy 1.19.5
Oauthlib 3.2.0
Onnx 1.10.2
Opencv - python 4.2.0.32
Openvino 2022.1.0
Opt - einsum 3.3.0
Pillow 6.2.0
PIP 21.3.1
The progress of 1.6
Protobuf 3.6.1 track
Py - cpuinfo 8.0.0
Pyasn1 0.4.8
Pyasn1 - modules 0.2.8
Pylibdmtx 0.1.9
Pyparsing 2.4.2
Python - dateutil 2.8.0
PyYAML 5.1.2
Requests 2.27.1
Requests - oauthlib 1.3.1
Rsa 4.8
Scipy 1.3.1
Setuptools 41.4.0
The six 1.15.0
Tensorboard 1.14.0
Tensorboard - data - server 0.6.1
Tensorboard plugin - wit, version 1.8.1
Tensorflow 1.14.0
Tensorflow - estimator 1.14.0
Tensorflow - model - optimization 0.1.3
Termcolor 1.1.0
The test - the generator while
TQDM 4.36.1
Typing - extensions 3.7.4.3
Urllib3 1.26.9
Virtualenv 16.7.5
Werkzeug 2.0.3
Wheel 0.37.1
Wrapt 1.12.1
The command for model transformation is
Py --input_model east_model_3t736_dm. pb --input_shape [1,736,736,3], --output east_detect/concat --data_type FP32/FP16,
The test input is the characteristic of [1,3,736,736], the expected output is [1,7,184,184], and the output is the offset relative to the original point set. The following is the correct output result of my FP16 model under FP16, which is also my expected output result:
This is my initial curing model and IR model under FP16
Here's the code I want to reproduce,
链接：https://pan.baidu.com/s/1RS3wXeNPZsN-SOz3io_DGw 
提取码：OTCV

**Comment by fanxiaochen456** [comment]
@jgespinoHello, THE OpenVINO version I'm using is OpenVino_2020.3.355 +VS2017+Python3.6.8. 
The command for model transformation is
Py --input_model east_model_3t736_dm. pb --input_shape [1,736,736,3], --output east_detect/concat --data_type FP32/FP16,
The test input is the characteristic of [1,3,736,736], the expected output is [1,7,184,184], and the output is the offset relative to the original point set. The following is the correct output result of my FP16 model under FP16, which is also my expected output result:
This is my initial curing model and IR model under FP16 and the OpenVINO environment configured and the graphics card I tested is as follows:
Here's the code I want to reproduce,
链接：https://pan.baidu.com/s/1uyE2uGXu9y1qeA3iuq8Btw 
提取码：OTCV

**Comment by fanxiaochen456** [comment]
@Iffa-MeahHello, this is my demo address, you can get the demo I want to test through the address, thank you.
https://github.com/fanxiaochen456/t.git

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 I tested your model.
The performance difference between CPU and GPU is around 0.76 FPS (which is halves).

Benchmark_app (CPU)
<img width="580" alt="12096_CPU" src="https://user-images.githubusercontent.com/79643476/178931791-65de71e3-23cf-4dcf-9e15-3b2c686806a4.png">

Benchmark_app (GPU)
<img width="583" alt="12096_GPU" src="https://user-images.githubusercontent.com/79643476/178931883-3492009c-47a7-4b75-a90e-008147a81172.png">


I observe the same behaviour as yours when I infer the model with CPU and GPU.
I used the official OpenVINO Classification Python Sample.

CPU produces a nice result
<img width="655" alt="12096_classificationCPU" src="https://user-images.githubusercontent.com/79643476/178932180-589b3d28-9654-4ca6-ae32-7dc8ef786f86.png">

Meanwhile the GPU produces NAN
<img width="658" alt="12096_classificationGPU" src="https://user-images.githubusercontent.com/79643476/178932333-25e1faee-74bf-45ca-8590-fbbc78094b74.png">

This is probably caused by the way you developed that model itself instead of OpenVINO.

The NAN indicates special values in DataFrame and arrays that represent the missing value in a cell.
This might relate to your cost function or is often referred to as a loss function. This function must be an average so that it can output a single value.

**Comment by fanxiaochen456** [comment]
@jgespino Hello, I have checked the loss function of this model, and it uses a custom loss function. So, besides replacing the loss function to retrain the model, is there any other way to replace the mode supported by OpenVINO?

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 not sure if you are actually trying to refer to me (replied to my answer above) or if you want @jgespino opinions about this since you mentioned him instead of me lol.

Anywhere here's my insight:

Compressing a model from its original format into FP16 would definitely have its drawback.
Most of the time the inferencing time gets better but the trade-off is accuracy (less accurate).
This might somewhat cause your issue too.

If your model has unsupported operations/functions, one of the possible workarounds is to create a custom operations/function which you may refer [here](https://docs.openvino.ai/latest/openvino_docs_Extensibility_UG_Intro.html#openvino-extensibility-mechanism).

Another option s to use the officially validated models that has similar implementation as yours.
You may refer to this [public](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/public) and [intel](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel) models.

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 as I explained above, there are a few factors that may lead to your issues:
1. Your model has unsupported operations/functions. (perhaps as you mentioned, your model uses a custom loss function which is not supported by OpenVINO)
2. Compressing a model from its original format or full precision FP32 into smaller in size precision (FP16/INT8) would definitely have its drawback. Most of the time the inferencing time gets better but the trade-off is accuracy (less accurate).

**Comment by fanxiaochen456** [comment]
@Iffa-Meah Hello, the solution you recommended mentioned that you can use C++ to customize the operation, right?

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 yes it is

**Comment by fanxiaochen456** [comment]
@Iffa-Meah Hello, I am not very good at C++, could you please provide some reference cases? It's about some custom layer processing cases and stuff like that.

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 you could refer to the guide I  provided above actually. ([OpenVINO Extensibility Mechanism](https://docs.openvino.ai/latest/openvino_docs_Extensibility_UG_Intro.html#openvino-extensibility-mechanism))
Please help to carefully read through the guide as it does provide the required codes.

**Comment by fanxiaochen456** [comment]
@Iffa-Meah Hello.When using C++ to carry out custom model reasoning, is it necessary to write some code supporting custom layer from the model of FP32 in C++?

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 necessary if you are using an[ unsupported layer with OpenVINO](https://docs.openvino.ai/latest/openvino_docs_MO_DG_prepare_model_Supported_Frameworks_Layers.html). If else, then no need.

**Comment by fanxiaochen456** [comment]
@Iffa-Meah Hello, my model contains a customized loss. According to the official website, should I use Cmake for recompilation?

**Comment by fanxiaochen456** [comment]
@Iffa-Meah @Iffa-Meah Hello, my model contains a customized loss. According to the official website, should I use Cmake for recompilation? I didn't find the core.add_extension function mentioned in the documentation.
![image](https://user-images.githubusercontent.com/101849755/179496362-39f934b5-fc30-499c-9ed0-f1faba72ec90.png)

**Comment by fanxiaochen456** [comment]
@Iffa-Meah Hello, can I refer to the example in text_detection_demo?

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 yes it's good to make an official samples/demos as reference as they are validated to work.
While you're at it, check on how the sample/demo handles the GPU too, and cross-check with yours.

**Comment by fanxiaochen456** [comment]
@Iffa-Meah Hello, I found the demo of text detection, but some packages involved in it could not be found and could not be implemented. For example, is there a tutorial for gflags.h?

**Comment by brmarkus** [comment]
For the "Google command line parser flags", you might want to check this guide: "https://gflags.github.io/gflags/".

**Comment by fanxiaochen456** [comment]
@brmarkus Thank you.
@Iffa-Meah Hello, if I change the model's custom loss function to MSE loss function after retraining, does OpenVINO support it?

**Comment by Iffa-Meah** [comment]
@fanxiaochen456 it should work as long as you are using a supported framework layer which you may refer [here](https://docs.openvino.ai/latest/openvino_docs_MO_DG_prepare_model_Supported_Frameworks_Layers.html).

But again, bear in mind (as I mentioned previously) **compressing** a model from its original format or full precision FP32 into smaller in size precision (FP16/INT8) **has drawback**. If your use case require to be accurate, such as clinical related results, it is not recommended to shrink the model as you need to bear with a less accurate prediction. Meanwhile, if your use case requires fast results without the need for prediction to be really accurate, then it is suitable to use a smaller size format.

[This](https://www.youtube.com/watch?v=RF8ypHyiKrY) might help you to understand better.

**Comment by fanxiaochen456** [comment]
@Iffa-Meah Hello, I changed the loss function for retraining after testing, and no NaN value appeared. The loss function will not participate in the model reasoning process. Why does the loss function of the original model affect the occurrence of NaN value in the reasoning process?

**Comment by Iffa-Meah** [comment]
Closing issue, feel free to re-open or start a new issue if additional assistance is needed.
